sim/state.ts

```diff
@@ -0,0 +1,340 @@
+		const state: /* Battle */ AnyObject = this.serialize(battle, BATTLE, battle);
+		state.field = this.serializeField(battle.field);
+		state.sides = new Array(battle.sides.length);
+		for (const [i, side] of battle.sides.entries()) {
+			state.sides[i] = this.serializeSide(side);
+		}
+		state.prng = battle.prng.seed;
+		state.hints = Array.from(battle.hints);
+		return state;
+	}
+
+	// Deserialization can only really be done on the root Battle object as
+	// the leaf nodes like Side or Pokemon contain backreferences to Battle
+	// but don't contain the information to fill it in because the cycles in
+	// the graph have been serialized as references. Once deserialzized, the
+	// Battle can than be restarted (and provided with a `send` function for
+	// receiving updates).
+	deserializeBattle(serialized: string | /* Battle */ AnyObject): Battle {
+		const state: /* Battle */ AnyObject =
+			typeof serialized === 'string' ? JSON.parse(serialized) : serialized;
+		const options = {
+			formatid: state.format,
+			seed: state.prngSeed,
+			rated: state.rated,
+			debug: state.debugMode,
+			// We need to tell that Battle we're creating that it's been
+			// deserialized so that it allows us to populate it correctly and
+			// doesn't attempt to start playing out until we're ready.
+			deserialized: true,
+			strictChoices: state.strictChoices,
+		};
+		for (const side of state.sides) {
+			const team = side.team.split(side.team.length > 9 ? ',' : '');
+			// @ts-ignore - index signature
+			options[side.id] = {
+				name: side.name,
+				avatar: side.avatar,
+				team: team.map((p: string) => side.pokemon[Number(p) - 1].set),
+			};
+		}
+		// We create the Battle, allowing it to instantiate the Field/Side/Pokemon
+		// objects for us. The objects it creates will be incorrect, but we descend
+		// down through the fields and repopulate all of the objects with the
+		// correct state.
+		const battle = new Battle(options);
+		this.deserialize(state, battle, BATTLE, battle);
+		this.deserializeField(state.field, battle.field);
+		for (const [i, side] of state.sides.entries()) {
+			this.deserializeSide(side, battle.sides[i]);
+		}
+		battle.prng = new PRNG(state.prng);
+		// @ts-ignore - readonly
+		battle.hints = new Set(battle.hints);
+		return battle;
+	}
+
+	serializeField(field: Field): /* Field */ AnyObject {
+		return this.serialize(field, FIELD, field.battle);
+	}
+
+	private deserializeField(state: /* Field */ AnyObject, field: Field) {
+		this.deserialize(state, field, FIELD, field.battle);
+	}
+
+	serializeSide(side: Side): /* Side */ AnyObject {
+		const state: /* Side */ AnyObject = this.serialize(side, SIDE, side.battle);
+		state.pokemon = new Array(side.pokemon.length);
+		const team = new Array(side.pokemon.length);
+		for (const [i, pokemon] of side.pokemon.entries()) {
+			state.pokemon[i] = this.serializePokemon(pokemon);
+			team[side.team.indexOf(pokemon.set)] = i + 1;
+		}
+		state.team = team.join(team.length > 9 ? ',' : '');
```

Comment:
This seems to be overcomplicating this field too.

I figure you are concerned about size of the serialized string? In that case we might as well omit ``.team`` from the output.

Reply:
No. I want to support 24 pokemon, so I was just encoding this as characters originally, but that led to confusion because index `a` (0) and `pokemon` `a` didn't line up. I chose 1-based number like the `team` command, and I don't like `1,2,3,4,5,6` because its too verbose. 

Basically, "i chose this format because it makes for the most readable output"

Reply:
Again, why would you "read" this output? To start with, JSON isn't meant to be human-readable. If we want readibility we should be using YAML.

Still, this is to be used to be parsed by a computer, so the simplest code is best.

Reply:
JSON is 100% meant to human readable. It's text! This is without a doubt one of the main reasons it is so popular and used in non-JS languages for protocol even though its clearly inferior bandwidth and parsing wise than a binary protocol. Binary is for computers, text is for humans. Albeit, in Javascript writing a binary protocol is a PITA, but you could still base64 encode the whole thing if it wasnt meant for humans.

I dont want to support YAML, YAML is a dependency and slower on JS, where JSON is native.

This is meant to be parsed by a computer, but I need to a) read it in order to debug it while I'm implementing it b) plan to continue reading it going forward. As I am working towards a MCTS based AI, I assume I'm going to be the primary user of this serialization, and having it readable is important to me.

---

`side.state.team` = the command *I* would need to issue at `/team` to transform `side.state.pokemon` into the `side.state.team` order. That is intuitively how I think about the meaning of the field, and after trying several different encoding schemes for team, it proves to be the simplest. The code to parse it is still trivial for the computer, and even if it werent, the size of a team is bounded so we could sacrifice performance for readability if we needed to.